### ISSUE TITLE
Typescript and createApi react usage

### ISSUE THREAD

**Issue by alexanderankin** [reduxjs/redux-toolkit#3268]
I am finding that e.g. for [this example in the docs](https://redux-toolkit.js.org/rtk-query/usage/migrating-to-rtk-query#api-slice-file), the resulting `useGetPokemonByNameQuery` variable is not actually present on `api`. So, you have to cast `api` to `any` to access it. and then the hook is `any`. but that hook should instead be a function of whatever `builder.query` defines as its second generic argument.

Maybe [this combination](https://github.com/reduxjs/redux-toolkit/blob/0734dad0061490d51755665d835c85ff3eb5cf12/packages/toolkit/src/query/react/index.ts#L19-L22) is using some new Typescript feature that I am failing to configure (not an expert user here (yet)). In which case I apologize for this issue at all.

Either way, maybe a couple hints for how to type annotated these hooks would be helpful.

**Comment by alexanderankin** [comment]
couldn't reproduce, restarting intellij and specifying the tsconfig.json here maybe helped?
![image](https://user-images.githubusercontent.com/8921892/225108788-27f07119-fdaa-48be-a178-46bf904c822b.png)

**Comment by alexanderankin** [comment]
also specifying EndpointDefinitions seems to break it as well. e.g.:

```typescript
export const authApiSlice = createApi({
  reducerPath: 'auth/api',
  baseQuery: fetchBaseQuery({
    baseUrl: '/abc',
  }),
  endpoints: (builder) :EndpointConfig /* here */ => ({
  }) //...
```

**Comment by markerikson** [comment]
FWIW, we've generally seen that Intellij IDEs seem to struggle more with TS in general, whereas VS Code works fine.

**Comment by phryneas** [comment]
Yeah, IntelliJ has rolled their own type interpreter and does not use the TypeScript language server. A lot of times, it's just wildly guessing based on things having similar names. We can impossibly support that.

But also, what Mark said: writing a type like `:EndpointConfig` erases important type information. Before it knew exactly what endpoints you had, after adding that it only knows that there may be endpoints.

**Comment by alexanderankin** [comment]
Thanks for the feedback. somehow getting similar messages in vscode and intellij, but this insight is helping me work things out. even if neither tool can fill in the blanks, i am somehow able to manually test (like if i put this here, string works, int doesn't).

**Comment by phryneas** [comment]
VSCode should perfectly support this out-of-the box with exactly the code in the example, as long as your TS is newer than the now-ancient 4.1

**Comment by phryneas** [comment]
See [this CodeSandbox as an example](https://codesandbox.io/s/pensive-visvesvaraya-lufzt4?file=/src/index.ts)

The only thing I can think of is that you should have `strict: true` in your `tsconfig.json`.
